File: AdventOfCode_2015/Day05/rust/src/bin/part1.rs
use itertools::Itertools;
use std::fs::File;
use std::io::{self, BufRead, BufReader};

fn is_nice(s: &str) -> bool {
    // At least three vowels
    let vowels = "aeiou";
    let vowel_count = s.chars().filter(|&c| vowels.contains(c)).count();
    if vowel_count < 3 {
        return false;
    }

    // At least one double letter
    if !s.chars().tuple_windows().any(|(a, b)| a == b) {
        return false;
    }

    // No forbidden substrings
    let forbidden = ["ab", "cd", "pq", "xy"];
    for &f in &forbidden {
        if s.contains(f) {
            return false;
        }
    }

    true
}

fn main() -> io::Result<()> {
    let path = "../input.txt";
    let file = File::open(path)?;
    let reader = BufReader::new(file);

    // let nice_count = reader
    //     .lines()
    //     .filter_map(|line| line.ok())
    //     .filter(|line| is_nice(line))
    //     .count();

    let nice_count = reader
        .lines() // This produces an iterator over Result<String, io::Error>
        .filter_map(|line| match line {
            Ok(l) => Some(l), // If line is Ok, continue
            Err(e) => {
                eprintln!("Error reading line: {}", e);
                None
            }
        })
        .filter(|line| is_nice(line)) // Filter nice strings
        .count(); // Count how many nice strings there are

    println!("Number of nice strings: {}", nice_count);

    Ok(())
}

File: AdventOfCode_2015/Day05/java/Part1.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;


public class Part1 {
    public static void main(String[] args) {
        String filepath = "../input.txt";

        try (BufferedReader reader = new BufferedReader(new FileReader(filepath))) {
            String line;
            int count = 0;
            while ((line = reader.readLine()) != null) {
                if (isNice(line)) {
                    count++;
                }
            }
            System.out.println(count);
        } catch (IOException e) {
            System.err.println("Error reading the input file: " + e.getMessage());
        }
    }

    private static boolean isNice(String line) {
        return hasTreeVowels(line) && hasDoubleLetter(line) && noForbiddenSubstrings(line);
    }

    private static boolean noForbiddenSubstrings(String line) {
        String[] forbidden = {"ab", "cd", "pq", "xy"};
        for (String sub : forbidden) {
            if (line.contains(sub)) {
                return false;
            }
        }
        return true;
    }

    private static boolean hasDoubleLetter(String line) {
        for (int i = 0; i < line.length() - 1; i++) {
            if (line.charAt(i) == line.charAt(i + 1)) {
                return true;
            }
        }
        return false;
    }

    private static boolean hasTreeVowels(String line) {
        String vowels = "aeiou";
        return line.chars().filter(c -> vowels.contains(String.valueOf((char) c))).count() >= 3;
    }


}
